src/evo/specialtx_validation.cpp

```diff
@@ -420,57 +420,79 @@ static bool CheckProUpRevTx(const CTransaction& tx, const CBlockIndex* pindexPre
 {
-    std::vector<CBLSPublicKey> allkeys;
-    for (const auto m : deterministicMNManager->GetAllQuorumMembers((Consensus::LLMQType)qfc.llmqType, pindexQuorum)) {
-        allkeys.emplace_back(m->pdmnState->pubKeyOperator.Get());
-    }
-    return qfc.Verify(allkeys);
-}
+    AssertLockHeld(cs_main);
 
-static bool CheckLLMQCommitmentTx(const CTransaction& tx, const CBlockIndex* pindexPrev, CValidationState& state)
-{
-    llmq::LLMQCommPL pl;
-    if (!GetTxPayload(tx, pl)) {
-        return state.DoS(100, false, REJECT_INVALID, "bad-qc-payload");
+    // Check version
+    if (qfc.nVersion == 0 || qfc.nVersion > llmq::CFinalCommitment::CURRENT_VERSION) {
+        return state.DoS(100, false, REJECT_INVALID, "bad-qc-quorum-version");
     }
 
-    if (pl.nVersion == 0 || pl.nVersion > llmq::LLMQCommPL::CURRENT_VERSION) {
-        return state.DoS(100, false, REJECT_INVALID, "bad-qc-version");
-    }
-
-    Optional<Consensus::LLMQParams> params = Params().GetConsensus().GetLLMQParams(pl.commitment.llmqType);
+    // Check type
+    Optional<Consensus::LLMQParams> params = Params().GetConsensus().GetLLMQParams(qfc.llmqType);
     if (params == nullopt) {
-        return state.DoS(100, false, REJECT_INVALID, "bad-qc-type");
+        return state.DoS(100, false, REJECT_INVALID, "bad-qc-quorum-type");
     }
 
-    if (!pl.commitment.VerifySizes(*params)) {
-        return state.DoS(100, false, REJECT_INVALID, "bad-qc-invalid-sizes");
+    // Check sizes
+    if (!qfc.VerifySizes(*params)) {
+        return state.DoS(100, false, REJECT_INVALID, "bad-qc-quorum-sizes");
     }
 
     if (pindexPrev) {
-        if (pl.nHeight != (uint32_t)pindexPrev->nHeight + 1) {
-            return state.DoS(100, false, REJECT_INVALID, "bad-qc-height");
-        }
-
-        if (!mapBlockIndex.count(pl.commitment.quorumHash)) {
+        // Get quorum index
+        auto it = mapBlockIndex.find(qfc.quorumHash);
+        if (it == mapBlockIndex.end()) {
             return state.DoS(100, false, REJECT_INVALID, "bad-qc-quorum-hash");
         }
-        const CBlockIndex* pindexQuorum = mapBlockIndex.at(pl.commitment.quorumHash);
+        const CBlockIndex* pindexQuorum = it->second;
         if (pindexQuorum != pindexPrev->GetAncestor(pindexQuorum->nHeight)) {
             // not part of active chain
             return state.DoS(100, false, REJECT_INVALID, "bad-qc-quorum-hash");
         }
 
-        if (!VerifyLLMQCommitment(pl.commitment, pindexQuorum)) {
-            return state.DoS(100, false, REJECT_INVALID, "bad-qc-invalid");
+        // Check height
+        if (pindexQuorum->nHeight % params->dkgInterval != 0) {
+            // not first block of DKG interval
+            return state.DoS(100, false, REJECT_INVALID, "bad-qc-quorum-height");
+        }
+
+        // Get members and check signatures (for not-null commitments)
+        std::vector<CBLSPublicKey> allkeys;
+        if (!qfc.IsNull()) {
```

Comment:
Can move `allkeys` inside the not null qfc block.

Reply:
Yep. Done.